### ISSUE TITLE
Make it possible to store some state together with the TaskInstance

### ISSUE THREAD

**Issue by tedsteen** [kagkarlsson/db-scheduler#5]
To be able to have some state associated with a task (without the need to keep a separate persistent store with task id -> state) it would be nice if one could store some state with a task instance.

#### Use case
I want to schedule an event to be published in the future.
Today I'd have to schedule a task and store the event together with the task id.
If I could store some state with my task I could put my event in that place and pick it up when the task executes.
```
scheduler.scheduleForExecution(triggerDateTime, new EventTask(taskName).instance(instanceId, "my state"));
....
@Override
public void execute(TaskInstance taskInstance, ExecutionContext executionContext) {
  //Do my thing with taskInstance.getState();
}
```

It could be a simple string and the user would have to handle serialization, or it could be more involved with serializers etc as part of the library.

**Comment by tedsteen** [comment]
An alternative, not to break the current api:
`new EventTask(taskName).instance(instanceId).withState("my state")`

Additions to TaskInstance:
```
private final String state;

private TaskInstance(Task task, String id, String state) {
	this.task = task;
	this.id = id;
	this.state = state;
}

public TaskInstance withState(String serializedState) {
	return new TaskInstance(task, id, serializedState);
}
```

**Comment by kagkarlsson** [comment]
Another good suggestion :). This feature has been pretty high on my todo-list as well. Primarily for the type of use-case you describe, to make one-time tasks more useful. The simplest thing would be to add a column to the table, an optional `state` field. Since it is optional, maybe we can add another method `instance(instanceId, state)`. 

I am unsure about what type to use for the state field.. need to think a bit about that. Would `byte[]` be slightly better than `String` perhaps, to encoding issues?
It's a goal to keep it simple, so maybe moving unmarshalling/deserialization to the "outside" would be a good thing. Just making it a bit harder to use.. 

Side-note: For extremely simple state, say a `long` primary key, it is possible to store it as the `instanceId´.

**Comment by tedsteen** [comment]
>I am unsure about what type to use for the state field.. need to think a bit about that. 

Either that, or use string and enforce a serializer that serializes to string to keep it simple.
There are also other [issues](http://stackoverflow.com/questions/3164072/large-objects-may-not-be-used-in-auto-commit-mode) when not using transactions in combination with binary large objects in postgre.

My suggested PR uses String. I'd say this is a simpler solution, but quite opinionated. All ears for suggestions tho :)

>Would byte[] be slightly better than String perhaps, to encoding issues?

Yes, or is InputStream a better API? (there is a `preparedStatement.setBlob(int, InputStream)`)

**Comment by tedsteen** [comment]
Did an alterative PR using byte[] as state. #8

**Comment by kagkarlsson** [comment]
Excellent, will try to have a look as soon as possible

**Comment by kagkarlsson** [comment]
Merged PR to master

**Comment by kagkarlsson** [comment]
Btw @tedsteen , when do you need a release with these new features?

**Comment by tedsteen** [comment]
There's no urgency here, I'll just use my fork in the meantime :)

**Comment by kagkarlsson** [comment]
Have a look at my branch `onetime-task-with-data` (commit e48af11c986415d3fab0fceb03ef56d6dfbfdf80) for an example on how you could get type-safe handling of the data. What do you think?

**Comment by tedsteen** [comment]
>Have a look at my branch onetime-task-with-data (commit e48af11) for an example on how you could get type-safe handling of the data. What do you think?

Thanks, yes that def makes sense. One thing I've been thinking about tho.
This is only applicable when scheduling. What about running persisted tasks after, f.ex a restart?
How do you make sure the correct implementation of execute is running? Where can I hook in to make sure my typesafe execute method is invoked when all due tasks are run after a restart?

**Comment by kagkarlsson** [comment]
The scheduler needs a reference to the task. It is looks for a task implementation in its registry matching the task-name from the database. From the example in the doc, notice the `create` method parameters

```
final MyAdhocTask myAdhocTask = new MyAdhocTask();

final Scheduler scheduler = Scheduler
        .create(dataSource, myAdhocTask)
        .threads(5)
        .build();

scheduler.start();

// Schedule the task for execution a certain time in the future
scheduler.scheduleForExecution(Instant.now().plusSeconds(5), myAdhocTask.instance("1045"));
```

Was this what you meant?

**Comment by tedsteen** [comment]
Yes, that cleared it up :) so that's what taskName is for. mapping implementations to database more or less?

**Comment by kagkarlsson** [comment]
Yeah, exactly :)

tir. 25. apr. 2017 kl. 11.38 skrev Ted Steen <notifications@github.com>:

**Comment by kagkarlsson** [comment]
This feature was implemented in version 2.0 thanks to the PR from @tedsteen .
Closing the issue.
